documentation/10_tls_with_exasol (#12)
exasol/tutorials: tls/doc/tls_with_exasol.md

1. The certificate is self-signed, so your client does not recognize it 
2. Exasol can’t predict where the certificate is going to be hosted, so the hostname is left blank in the certificate

### The Chicken-or-the-egg Problem of Secure Connections

For a truly secure connection, you need to put in place a custom server certificate immediately after installing the cluster. Until then, you’re facing a classic chicken-or-the-egg problem. On the one hand, you need to install your own certificate to establish a trusted TLS connection. On the other hand, you need to connect to Exasol first to do this.

The graphic below shows an example of a simple on-premise installation in a single network. The organization’s CA issues a root certificate. IT administrators then place this certificate on all users’ machines in that organization. The same administrators also install the server certificate. There are multiple options, but in our example they do it via an SSH connection.

![Installing a server certificate](images/seq_installing_a_server_certificate.svg)

### Exasol TLS Certificate Installation (On-premise)

> You can be reasonably sure with on-premise that it takes a matter of minutes to replace the certificate — and that no one will interfere.

Reviewer: ```suggestion
With on-premise setups, you can be reasonably sure that replacing the certificate takes only a few minutes — and that no one will interfere.
```

## Outgoing Connections

Depending on the certificate issuer, outgoing TLS connections can either be as easy as pie or a right royal pain.

### Services With Certificates Issued by a Popular CA

If an application on your Exasol cluster wants to connect to the public application programming interface (API) of a
service, there is little to do if that service uses a certificate signed by a popular CA. These are the CAs that come
preloaded in most truststores. The TLS connection should work out of the box in this case. The TLS client can verify the
server certificate by following the certificate chain down to that known root CA.

### Services With Certificates Issued by a Custom CA

When you want to connect Exasol to a service that has a certificate issued by your own CA, you need to work a bit harder to get TLS running. There are multiple reasons:

1. If you run an Exasol cluster, all data nodes must know the custom root CA — the certificate must be installed on all nodes 
2. [User Defined Functions (UDFs) can’t access the standard truststore](tls_in_udfs.md) as they run in an environment that only gives them very restricted file system access 

> 3. The truststores the UDF container sees by default are installed with the [Script Language Containers (SLCs)](https://github.com/exasol/script-languages-release) that means that getting in different certificates means either building custom language containers or completely overriding the truststore

Reviewer: ```suggestion
3. The truststores visible to the UDF container are those installed with the [Script Language Containers (SLCs)](https://github.com/exasol/script-languages-release) that means that getting in different certificates means either building custom language containers or completely overriding the truststore
```

* You can write a shell script and loop through the hosts.
* Up to Exasol 7.1 you can upload certificates through ExaOperation

Please refer to the section "[Upload TLS Certificate](https://docs.exasol.com/db/latest/administration/on-premise/access_management/tls_certificate.htm)" in Exasol's online handbook for recommended practices.

#### Using TLS in User Defined Functions (UDFs)

UDFs run in a sandbox for security reasons. The main effect you should be aware of in the context of TLS is that this means UDFs cannot see the host filesystem. And that means that the certificates on the host filesystem are visible to the EXALoader, but **not** to UDFs.

Since this is a tricky topic, we have a separate article about using [TLS in UDFs](tls_in_udfs.md).

### Combining UDFs With the EXALoader

This is the point where things get really complicated. As mentioned before, UDFs run in a sandbox. So they cannot see the certificate files on the host filesystem. The EXALoader can, though.

In a scenario where you need both — let's say you want to use a Virtual Schema — you have different options.

> Remember that if UDF and EXALoader only need to connect to servers that use a TLS certificate signed by a CA that is in the default trust stores, you don't have to do anything.

Reviewer: ```suggestion
Remember, if the UDF and EXALoader only need to connect to servers using TLS certificates signed by a CA already included in the default truststores, no additional action is required.
```
